### ISSUE TITLE
Unused Closure Parameter "self" but "self" was actually used in closure

### ISSUE THREAD

**Issue by hisoka0917** [realm/SwiftLint#2349]
SwiftLint give a warning here "self" is not used. But "self" was actually used in the closure.

![972352ad-28a9-4e90-a3b4-3d17f83de2bf](https://user-images.githubusercontent.com/34638425/43882485-5deced78-9be2-11e8-975e-d37a4bbedc2e.png)

I use a pod named `Delegated` that allow me not write [weak self]. Here is the implement of delegate:

```
public mutating func delegate<Target : AnyObject>(to target: Target,
                                                      with callback: @escaping (Target, Input) -> Output) {
        self.callback = { [weak target] input in
            guard let target = target else {
                return nil
            }
            return callback(target, input)
        }
    }
```
SwiftLint version is 0.27.0
Xcode version is 9.2

Is it a bug?

**Comment by marcelofabri** [comment]
@hisoka0917 I copied this snippet and it didn't trigger that violation:

```
➜ swiftlint lint --path file.swift
Loading configuration from '.swiftlint.yml'
Linting Swift files at paths file.swift
Linting 'file.swift' (1/1)
file.swift:1:31: warning: Colon Violation: Colons should be next to the identifier when specifying a type and next to the key in dictionary literals. (colon)
Done linting! Found 1 violation, 0 serious in 1 file.
➜  cat file.swift 
public mutating func delegate<Target : AnyObject>(to target: Target,
                                                  with callback: @escaping (Target, Input) -> Output) {
    self.callback = { [weak target] input in
        guard let target = target else {
            return nil
        }
        return callback(target, input)
    }
}
```

**Comment by hisoka0917** [comment]
@marcelofabri Let me show you a full example.

`Delegated` is a pod library. Add `pod 'Delegated', '~> 0.1.0'` to `Podfile`. Or you can just add this file to your project.

And because it's a pod, so it was excluded in `.swiftlint.yml`.

Then use it. First I define a class: 

```
import Delegated

class MyObject {
    var handler = Delegated<Void?, Void>()
}

class MyClass {

    func doSomething() {
        print("do something")
    }

    func callDelegate() {
        let object = MyObject()
        object.handler.delegate(to: self) { (self, _) in
            self.doSomething()
        }
    }
}
```
The warning occurs in this line `object.handler.delegate(to: self) { (self, _) in`.

<img width="1022" alt="2018-08-09 11 31 14" src="https://user-images.githubusercontent.com/34638425/43909247-7158eb10-9c2c-11e8-8c46-3845cb8dab22.png">

`Delegated` is very simple, you can see the implement here: [Delegated](https://github.com/dreymonde/Delegated/blob/master/Sources/Delegated/Delegated.swift)
I don't think there is something wrong in `Delegated`. Otherwise I think it's a bug in SwiftLint.

**Comment by marcelofabri** [comment]
@hisoka0917 this is because `self` gets recognized as a keyword, not an identifier by SourceKit.

Renaming the variable to `delegate` for example fixes the issue. If you think it's important to be able to use `self`, I'd open an issue on https://bugs.swift.org asking if the inner `self` shouldn't be a recognized as `source.lang.swift.syntaxtype.identifier` instead of `source.lang.swift.syntaxtype.keyword`.

**Comment by hisoka0917** [comment]
@marcelofabri I think it's useful. Since we usually need to avoid retain cycle, so we often write code like this: 

```
object.handler = { [weak self] _ in 
    if let strongSelf = self {
        strongSelf.doSomething()    
    }
}
```

Codes like this is everywhere and it's not elegant, is it? 

So I need something like `Delegated` to avoid `[weak self]`. And I think there must be somewhere else should use `self` as identifier.

I would be very grateful if open an issue on `swift.org`. Thank you.

**Comment by marcelofabri** [comment]
Well, you can still use the same thing but just using a different variable name than `self`...

```swift
object.handler.delegate(to: self) { (me, _) in
    me.doSomething()
}
```
